WIP: iptables firewall rules

chriskuehl opened the pull request:
This adds two new modules:

* `paasta_tools.iptables` which contains helper functions for working with iptables. It doesn't contain any knowledge of PaaSTA or services.
* `paasta_tools.firewall` which sets up the PaaSTA chains and rules. Currently it doesn't know how to list running services or what their rules should be (those functions just return fake data), but the update script does the right adds/modifies/deletes based on that fake data. (This module will need to be merged with the firewall module bchess recently added.)

Currently the interface is a single "update" script. Below are some examples of what it looks like currently (produced by modifying the functions that return the lists of running services). After each scenario, running again is a noop.

We'll need to figure out the integration points between this and launching/stopping containers. This code as written is pretty much ready to go into a cronjob, except for the fake data bits.

## Examples
#### Starting with a fresh machine (no iptables rules, no services running)

```
$ sudo python2.7 -m paasta_tools.firewall                                  
creating chain: PAASTA-INTERNET
adding rule to PAASTA-INTERNET: -j ACCEPT
adding rule to PAASTA-INTERNET: -d 127.0.0.0/8 -j RETURN
adding rule to PAASTA-INTERNET: -d 10.0.0.0/8 -j RETURN
adding rule to PAASTA-INTERNET: -d 172.16.0.0/12 -j RETURN
adding rule to PAASTA-INTERNET: -d 192.168.0.0/16 -j RETURN
adding rule to PAASTA-INTERNET: -d 169.254.0.0/16 -j RETURN
creating chain: PAASTA
adding rule to INPUT: -j PAASTA
adding rule to FORWARD: -j PAASTA
```

These are the base rules that are _always_ present on PaaSTA boxes.

#### Launching a couple services

```
$ sudo python2.7 -m paasta_tools.firewall
creating chain: PAASTA.cool-servi.ccd68729b4
adding rule to PAASTA.cool-servi.ccd68729b4: -j REJECT --reject-with icmp-port-unreachable
adding rule to PAASTA.cool-servi.ccd68729b4: -d 169.254.255.254/32 -p tcp -m tcp --dport 20666 -j ACCEPT
creating chain: PAASTA.cool-servi.003bae64d2
adding rule to PAASTA.cool-servi.003bae64d2: -j REJECT --reject-with icmp-port-unreachable
adding rule to PAASTA.cool-servi.003bae64d2: -d 169.254.255.254/32 -p tcp -m tcp --dport 20666 -j ACCEPT
creating chain: PAASTA.dumb-servi.bcc6a6f4b1
adding rule to PAASTA.dumb-servi.bcc6a6f4b1: -j REJECT --reject-with icmp-port-unreachable
adding rule to PAASTA.dumb-servi.bcc6a6f4b1: -d 169.254.255.254/32 -p tcp -m tcp --dport 20666 -j ACCEPT
adding rule to PAASTA: -m mac --mac-source FE:A3:A3:DA:2D:40 -j PAASTA.cool-servi.ccd68729b4
adding rule to PAASTA: -m mac --mac-source 02:42:A9:FE:00:02 -j PAASTA.cool-servi.003bae64d2
adding rule to PAASTA: -m mac --mac-source FE:A3:A3:DA:2D:50 -j PAASTA.cool-servi.003bae64d2
adding rule to PAASTA: -m mac --mac-source FE:A3:A3:DA:2D:51 -j PAASTA.cool-servi.003bae64d2
adding rule to PAASTA: -m mac --mac-source FE:A3:A3:DA:2D:30 -j PAASTA.dumb-servi.bcc6a6f4b1
adding rule to PAASTA: -m mac --mac-source FE:A3:A3:DA:2D:31 -j PAASTA.dumb-servi.bcc6a6f4b1
```

#### Stopping one of those services

```
$ sudo python2.7 -m paasta_tools.firewall
deleting rule from PAASTA: -m mac --mac-source FE:A3:A3:DA:2D:40 -j PAASTA.cool-servi.ccd68729b4
deleting chain: PAASTA.cool-servi.ccd68729b4
```

chriskuehl committed:
Add some simple iptables managing functions

The most important of these is `ensure_chain` for creating and managing
the rules on an entire chain.

chriskuehl committed:
Add some logic for managing PaaSTA-specific rules

bchess commented on paasta_tools/firewall.py:
```diff
@@ -0,0 +1,117 @@
+from paasta_tools import iptables
+
+
+PRIVATE_IP_RANGES = (
```
There probably are packages out there where these constants are defined, but I prefer what you've done here and made it explicit and under our control. ++

bchess commented on paasta_tools/iptables.py:
```diff
@@ -0,0 +1,113 @@
+    output = subprocess.check_output(
+        ('iptables', '-t', 'filter', '--list-rules'),
+    )
+    chain_regex = re.compile('-N ([^ ]+)$')
```
pull chain_regex out to module level

bchess commented on paasta_tools/iptables.py:
```diff
@@ -0,0 +1,113 @@
+
+Unlike the `firewall` module, these functions know nothing about PaaSTA and
+could effectively be a third-party library. They just make working with
+iptables a little bit easier.
```
Did you do any search/evaluation for existing iptables packages on pypi?

chriskuehl replied:
I didn't originally, but I did just now. Of the ones that look to be maintained/reasonably popular, the one that looks most interesting to me is https://github.com/ldx/python-iptables.

Unlike my code, it provides Python bindings for working with the C interface to iptables. That would almost certainly be more performant and may make some things easier (e.g. maybe we no longer have to worry about parsing the rules, since it provides access to the structs used by the C API).

I'll update my review with the changes so far, and look into using this library.

bchess replied:
Great if it works. Yeah I was mostly thinking about something that'd be able to use a code interface rather than the iptables executable, so great you found one.

bchess commented on paasta_tools/iptables.py:
```diff
@@ -0,0 +1,113 @@
+        create_chain(chain)
+        current_rules = set()
+
+    if rules is not None:
```
so if rules is None, then no existing rules are removed? Is that needed? At the very least should go in docstring

chriskuehl replied:
yeah, originally i was using this to ensure a chain existed but not manage the rules, but I changed the approach to not require that (at least not for now).

I'll just delete that special behavior since we're not currently using it.

bchess commented on paasta_tools/iptables.py:
```diff
@@ -0,0 +1,113 @@
+        add_rule(chain, rule)
+
+
+def add_rule(chain, rule):
```
A bit bikesheddy, but imho since we're putting these rules at the top of the chain rather than appending, call this insert_rule() instead of add_rule()

bchess commented on paasta_tools/firewall.py:
```diff
@@ -0,0 +1,117 @@
+    """A service group.
+
+    :param service: the name of a service
+    :param dependency_group: the name of a dependency group
```
What's a dependency group?

bchess replied:
Oh, it's the name of the key in the dependencies.yaml dict

chriskuehl replied:
i'll make the docstring more clear

bchess commented on paasta_tools/iptables.py:
```diff
@@ -0,0 +1,113 @@
+                add_rule(chain, rule)
+
+        for rule in current_rules:
+            if rule not in rules:
```
Just mentally modeling this -- current_rules is a set of rulestrings, whereas rules is a tuple of rulestrings. So the first `__contains__` is O(1) and this second is O(n). You could build a set() out of rules for the deletion phase to improve that.

I'm a bit nervous about the direct string comparisons. Should we do some parsing/formatting of rulestrings? OTOH, it does look like `iptables list` orders the parameters of each rule consistently...

chriskuehl replied:
Yeah, that mental model is right. Just to be explicit, the reason for that is that when inserting rules, the insertion order actually matters (the REJECT needs to be inserted first). Turning it into a set during the second stage sounds like a good idea, I'll do that.

iptables definitely normalizes the rules when they are printed. That's also why I've written some of the rules in a slightly strange way. For example if you define a rule `-A INPUT -p tcp --dport 8888  -j REJECT`, it actually creates a rule that says `-A INPUT -p tcp -m tcp --dport 8888 -j REJECT --reject-with icmp-port-unreachable`.

My main concern would be if iptables changes the normalization, which could conceivably happen during e.g. an OS upgrade. In that case, this code would see the right rules but with the wrong normalization and remove/re-add them on every run. I don't have a great solution for that besides parsing the rules. Does that seem like a good idea to you?

chriskuehl commented on paasta_tools/firewall.py:
```diff
@@ -0,0 +1,117 @@
+
+
+def general_update():
+    """Update iptables to match the current PaaSTA state."""
```
One other thing to mention is that this is currently one function that does the entire update. To start, it's probably okay to call this every time there's a new service (or a service exits), but eventually we will probably need to extend the interface (to add functions for e.g. new instances or exiting instances) for performance reasons.

The code is fast now, but probably not fast enough for hundreds of instances, especially once it starts doing slow things like parsing yaml.

chriskuehl committed:
Address review feedback

bchess commented:
looks good so far!

chriskuehl committed:
Switch to using iptc

chriskuehl commented:
@bchess added another draft, now using iptc.

Some remaining TODOs:
* A few things could be more efficient. Adding rules is _really_ fast, but deleting is extremely slow when there are lots of rules (since it lists every rule on the chain and compares, per each rule to be deleted). It should be possible to make that way faster.
* It's a little verbose when constructing `Rule` objects, maybe it should have defaults
* Probably needs a bit more testing

bchess commented on paasta_tools/iptables.py:
```diff
@@ -0,0 +1,145 @@
+    print('deleting rule from {}: {}'.format(chain_name, rule))
+    chain = iptc.Chain(iptc.Table(iptc.Table.FILTER), chain_name)
+    for potential_rule in chain.rules:
+        if Rule.from_iptc(potential_rule) == rule:
```
Hmm. Yeah this is unfortunate. :(

bchess replied:
did you see the Autocommit stuff of iptc? This pattern seems to be directly addressed in that section of the README.

chriskuehl replied:
Yeah, I was using it here originally but temporarily deleted it. I'll need to figure out how to refactor this function to be fast anyway, just doing it without autocommit won't actually speed things up. Not using autocommit is more of a correctness thing, if I'm reading correctly their docs.

chriskuehl committed:
Turn off autocommit when deleting rules

This isn't a performance improvement, but it is a correctness
improvement.

chriskuehl committed:
Bulk delete rules

This greatly improves speed when deleting lots of rules from a chain.
It's still somewhat slow to delete a small number of rules from a large
chain, though.

bchess commented on paasta_tools/iptables.py:
```diff
@@ -0,0 +1,161 @@
+    https://github.com/ldx/python-iptables#autocommit
+    """
+    try:
+        table.autocommit = False
```
assert table.autocommit == True first

bchess commented:
overall lg2m

chriskuehl commented:
cool, I'm going to close this and integrate this into your paasta_tools changes, then open a PR upstream
